plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1371,6 +1378,141 @@ \subsubsection{Comparison of the typed CK machine and the untyped CEK machine}
+    encoding.
+  \item \texttt{Unit} and data structures which have only one constructor, are removed
+    from the serialised data.
+  \item Variable names are encoded using DeBruijn indices, which are unsigned, arbitrary
+    length integers.
+\end{itemize}
+
+Encoded values are aligned to byte/word boundary using a meaningless sequence of `0` bits
+terminated with a `1` bit.
+
+\subsection{Constants}
+
+Constants are encoded as a combination of the kind of value that is serialized and the
+value itself. Constants use 4 bits to encode the constructors, so they allow for a
+maximum of 16 constructors, of which 6 are currently used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|l|c|l|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Encoding} \\
+  \hline
+  ByteString & 1 & Variable length \rule{0mm}{4mm} \\[\sep]
+  Integer & 0 & ZigZag + Variable length \\[\sep]
+  String & 2 & UTF-8 \\[\sep]
+  Char & 3 & UTF-8 \\[\sep]
+  Unit & 4 & Empty \\[\sep]
+  Bool & 5 & `1` is True, `0` is False \\[\sep]
+  \hline
+\end{tabular}
+\captionof{figure}{Constant kinds}
+\label{fig:serialisation-constants}
+\end{minipage}
+
+\subsection{Untyped terms}
+
+Terms are encoded using 4 bits, which allows a total of 16 terms, of which 8 are
+currently used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|l|c|l|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Arguments} \\
+  \hline
+  Variable & 0 & name \rule{0mm}{4mm} \\[\sep]
+  Delay & 1 & term \\[\sep]
+  Lambda abstraction & 2 & name, term \\[\sep]
+  Application & 3 & term, term \\[\sep]
+  Constant & 4 & constant \\[\sep]
+  Force & 5 & term \\[\sep]
+  Error & 6 & term \\[\sep]
+  Builtin & 7 & builtin \\[\sep]
+  \hline
+\end{tabular}
+\captionof{figure}{Untyped terms}
+\label{fig:serialisation-terms}
+\end{minipage}
+
+\vspace{1cm}
+
+\noindent Builtins use 5 bits for their definition, allowing for a maximum of 32 builtin
+functions of which 25 are used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|c|c|c|c|c|c|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} \\
+  \hline
+   AddInteger & 0 & GreaterThanEqInteger & 8 & QuotientInteger & 17 \rule{0mm}{4mm} \\[\sep]
+   SubtractInteger & 1 & EqInteger & 9 & ModInteger & 18 \\[\sep]
+   MultiplyInteger & 2 & Concatenate & 10 & LtByteString & 19 \\[\sep]
+   DivideInteger & 3 & TakeByteString & 11 & GtByteString & 20 \\[\sep]
+   RemainderInteger & 4 & DropByteString & 12 & IfThenElse & 21 \\[\sep]
+   LessThanInteger & 5 & SHA2 & 13 & CharToString & 22 \\[\sep]
+   LessThanEqInteger & 6 & SHA3 & 14 & Append & 23 \\[\sep]
+   GreaterThanInteger & 7 & VerifySignature & 15 & Trace & 24 \\[\sep]
+   & & EqByteString & 16 & & \\[\sep]
+   \hline
+\end{tabular}
+\captionof{figure}{Builtin tags}
+\label{fig:serialisation-builtins}
+\end{minipage}
+
+\vspace{1cm}
+
```

Comment:
Do you think I should include the `Program` definition in the specification?

Reply:
There's an interesting question here. What people often do is have a more-stable "container" format that is basically just a version tag so you can tell how to interpret the body. We could do the same here, and specify the version header as the first few bytes of the file, and state that this is intended to be stable even if the body encoding changes later.

I like this, it would let us be fairly confident that we could always deserialize PLC files without having to know what version they are beforehand.

Reply:
We always encode a `Program` which does contain versioning information, so I think we are already doing that?

Reply:
Well:
a) We want to be *extra* sure that the outer encoding is stable, so it will *always* be three variable length integers in exactly the same format, no matter what else changes. So that's worth calling out separately.
b) It's easy for that to go wrong if, say, we ever make use of programs with non-unit annotations (which we might want for, say, debugging). Then the annotation for Programs would go before the version, oh no!

So I think we need a specification of the header separately, and maybe a custom serialization format for Programs that puts the version before the annotation (and maybe remove the annotation for `Version`, too).

Reply:
I was wondering how you'd deal with scripts that are using non-standard builtins, since then the encoding of a builtin could refer to something completely unexpected.  I think we discussed this before and decided that it was just up to users to make sure they were doing the right thing.  Now I'm wondering if we could generate a hash form the names of the types and the signatures of the builtin functions and stick it in the header somewhere.